### PR TITLE
Laboratory work #3, Shuvar Alena - 19FPL2

### PR DESCRIPTION
**Pull request by alenashuvar**

### CODE REVIEW

File: lab_3/main.py [hunk 1]
```diff
@@ -17,38 +19,62 @@ def tokenize_by_sentence(text: str) -> tuple:
...
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(letter, str) or not 0 < len(letter) <= 1:
+            return 1
+        if letter not in self.storage:
+            self.storage[letter] = len(self.storage) + 1
```
Comment (WhiteJaeger) [inline lab_3/main.py:52]: Great!

File: lab_3/main.py [hunk 2]
```diff
@@ -59,49 +85,98 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
...
     :return: a tuple of the encoded sentences
     """
-    pass
+    if not isinstance(storage, LetterStorage) or not isinstance(corpus, tuple):
+        return ()
+
+    encoded_corpus = []
+    for sentence in corpus:
+        list_sentence = []
+        if isinstance(sentence[0], tuple):
```
Comment (WhiteJaeger) [inline lab_3/main.py]: why is this check needed?

File: lab_3/main.py [hunk 3]
```diff
@@ -119,15 +206,47 @@ def _calculate_distance(self, first_n_grams: tuple, second_n_grams: tuple) -> in
         :param second_n_grams: a tuple of the top_k n-grams
         :return: a distance
         """
-        pass
+        incorrect_inputs = (not isinstance(first_n_grams, tuple) or
+                            not isinstance(second_n_grams, tuple) or
+                            len(first_n_grams) and not isinstance(first_n_grams[0], (tuple, str)) or
+                            len(second_n_grams) and not isinstance(second_n_grams[0], (tuple, str)))
```
Comment (WhiteJaeger) [inline lab_3/main.py]: Could it not be shortened?
Without `len(second_n_grams)`\ `len(first_n_grams)`

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 2 above)
Comment (alenashuvar) [reply lab_3/main.py]: I couldn't pass all the tests without it, but I pulled updates yesterday, so now all is working and I will remove it

File: lab_3/main.py [hunk 4]
```diff
@@ -17,38 +19,62 @@ def tokenize_by_sentence(text: str) -> tuple:
...
-    pass
+    if not isinstance(text, str) or len(text) == 0:
+        return ()
+
+    sentences = re.split('[!?.] ', text)
+    list_letters = []
+
+    for sentence in sentences:
+        list_tokens = re.sub('[^a-z \n]', '', sentence.lower()).split()
+        if len(list_tokens) == 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py]: `len(list_tokens) == 0` can be shortened.

File: lab_3/main.py [hunk 5]
```diff
@@ -59,49 +85,95 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
...
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if len(self.n_grams) == 0:
+            return 1
+
+        for sentence in self.n_grams:
+            for word in sentence:
+                for n_gram in word:
+                    self.n_gram_frequencies[n_gram] = self.n_gram_frequencies.get(n_gram, 0) + 1
```
Comment (WhiteJaeger) [inline lab_3/main.py:139]: Good!

File: lab_3/main.py [hunk 6]
```diff
@@ -59,49 +85,95 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
...
+        return 0
 
     def calculate_log_probabilities(self) -> int:
         """
         Gets log-probabilities of n-grams, fills the field n_gram_log_probabilities
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if len(self.n_gram_frequencies) == 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py]: `len(self.n_gram_frequencies) == 0` can be shortened.

File: lab_3/main.py [hunk 7]
```diff
@@ -59,49 +85,95 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
...
+            self.n_gram_log_probabilities[n_gram] = log(probability)
+        return 0
 
     def top_n_grams(self, k: int) -> tuple:
         """
         Gets k most common n-grams
         :return: a tuple with k most common n-grams
         """
-        pass
+        if not isinstance(k, int) or k < 0 or len(self.n_gram_frequencies) == 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py]: `len(self.n_gram_frequencies) == 0` can be shortened.

File: lab_3/main.py [hunk 8]
```diff
@@ -110,42 +182,104 @@ def new_language(self, encoded_text: tuple, language_name: str) -> int:
...
         Calculates distance between top_k n-grams
         :param first_n_grams: a tuple of the top_k n-grams
         :param second_n_grams: a tuple of the top_k n-grams
         :return: a distance
         """
-        pass
+        incorrect_inputs = (not isinstance(first_n_grams, tuple) or
+                            not isinstance(second_n_grams, tuple) or
+                            None in first_n_grams or
+                            None in second_n_grams)
```
Comment (WhiteJaeger) [inline lab_3/main.py]: It is too specific, try to generalize this check.
You should check if the input args are _appropriate_.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]
Awesome work! Fix some minor issues and you are done here.

File: lab_3/main.py [hunk 9]
```diff
@@ -17,38 +19,62 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
-    pass
+    if not isinstance(text, str) or len(text) == 0:
```
Comment (WhiteJaeger) [inline lab_3/main.py]: `len(text) == 0` can be shortened.

File: lab_3/main.py [hunk 10]
```diff
@@ -17,38 +19,62 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        self.storage = {}
 
     def _put_letter(self, letter: str) -> int:
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(letter, str) or not 0 < len(letter) <= 1:
```
Comment (WhiteJaeger) [inline lab_3/main.py]: `0 <` is redundant.

**Review by WhiteJaeger — APPROVED** [review]
Great work! You are done here. Please, do not forget to fix comments.

**Review by WhiteJaeger — APPROVED** [review]

**Pull request merged**